nimbus/src/lib.rs

```diff
@@ -138,6 +138,42 @@ impl NimbusClient {
         Ok(events)
     }
 
+    /// Callback to signal application-level telemetry opt-out.
+    ///
+    /// When the user opts out of telemetry in the consuming application, we need to reset any
+    /// unique identifiers used internally by the SDK. If we don't then we risk accidentally
+    /// tracking users who out out of telemetry and opt back in, since we could use Nimbus metrics
+    /// to link their telemetry pings from before they opted out, with their pings from after
+    /// they opted back in.
+    ///
+    /// This method must uphold two important properties in order to work properly for consumers:
+    ///   * it must be idempotent, so that consumers can safely call it multiple times rather
+    ///     than trying to remember whether or not it was successfully called.
+    ///   * it must not change the enrolled branch of any active experiments, so that opting out
+    ///     of telemetry doesn't cause unrelated changes to other parts of the application UI.
+    ///
+    pub fn on_telemetry_disabled(&mut self) -> Result<()> {
+        let db = self.db()?;
+        let mut writer = db.write()?;
+        // If we have no `nimbus_id` when we can safely assume that there's
+        // no other experiment state that needs to be reset.
+        let store = db.get_store(StoreId::Meta);
+        Ok(
+            if let Some(_) = store.get::<String>(&writer, DB_KEY_NIMBUS_ID)? {
+                // Each enrollment state includes a unique `enrollment_id`.
+                // TODO: Ideally, we would wipe existing experiment enrollments at the next
+                // available opportunity. We don't do it immediately because we don't want
+                // to trigger weird changes in application UI.
+                // We could add to the "pending changes" infra from https://github.com/mozilla/nimbus-sdk/pull/78
+                // once it lands.
+                enrollment::on_telemetry_disabled(db, &mut writer)?;
```

Comment:
If we do end up enqueuing a "wipe enrollments" flag to be applied at the next startup, we may not need to reset the enrollment identifiers at all.